circuit/src/utils.rs

```diff
@@ -26,7 +26,104 @@ use halo2::{
 };
 use num_bigint::BigUint;
 use rand::Rng;
-use std::{fmt::Debug, fs::write, io::Read, time::Instant};
+use serde::{de::DeserializeOwned, Serialize};
+use std::{
+	env::current_dir,
+	fmt::Debug,
+	fs::{write, File},
+	io::{BufReader, Error as IoError, Read},
+	path::Path,
+	time::Instant,
+};
+
+/// Reads raw bytes from the file
+pub fn read_bytes(path: impl AsRef<Path>) -> Vec<u8> {
+	let f = File::open(path).unwrap();
+	let mut reader = BufReader::new(f);
+	let mut buffer = Vec::new();
+
+	// Read file into vector.
+	reader.read_to_end(&mut buffer).unwrap();
+
+	buffer
+}
+
+/// Reads string from the file
+pub fn read_string(path: impl AsRef<Path>) -> String {
+	let f = File::open(path).unwrap();
+	let mut reader = BufReader::new(f);
+	let mut buffer = String::new();
+
+	// Read file into string.
+	reader.read_to_string(&mut buffer).unwrap();
+
+	buffer
+}
+
+/// Reads raw bytes from the file
+pub fn read_bytes_data(name: &str) -> Vec<u8> {
+	let current_dir = current_dir().unwrap();
+	let bin_path = current_dir.join(format!("../data/{}.bin", name));
+	read_bytes(bin_path)
+}
+
+/// Write bytes to a file
+pub fn write_bytes(bytes: Vec<u8>, path: impl AsRef<Path>) -> Result<(), IoError> {
+	write(path, bytes)
+}
+
+/// Write bytes to data directory
+pub fn write_bytes_data(bytes: Vec<u8>, name: &str) -> Result<(), IoError> {
```

Comment:
Same as above.
I think it is better to use clear variable name, like `file_name`, instead of just `name`.